File: plugin/src/shared/sc/plugin2020/util/Direction.java
package sc.plugin2020.util;

public enum Direction {
  RIGHT,
  LEFT,
  UPRIGHT,
  UPLEFT,
  DOWNRIGHT,
  DOWNLEFT;

  public CubeCoordinates shift(int d) { // d is the distance
    int shiftX = 0;
    int shiftY = 0;
    int shiftZ = 0;
    switch (this) {
      case RIGHT:
        shiftX = 1*d;
        shiftY = -1*d;
        break;
      case LEFT:
        shiftX = -1*d;
        shiftY = 1*d;
        break;
      case UPRIGHT:
        shiftX = 1*d;
        shiftZ = -1*d;
        break;
      case UPLEFT:
        shiftY = 1*d;
        shiftZ = -1*d;
        break;
      case DOWNRIGHT:
        shiftY = -1*d;
        shiftZ = 1*d;
        break;
      case DOWNLEFT:
        shiftX = -1*d;
        shiftZ = 1*d;
        break;
    }
    return new CubeCoordinates(shiftX, shiftY, shiftZ);
  }
}

File: plugin/src/shared/sc/plugin2020/util/Direction.kt
package sc.plugin2020.util

enum class Direction {
    RIGHT,
    LEFT,
    UPRIGHT,
    UPLEFT,
    DOWNRIGHT,
    DOWNLEFT;

    fun shift(distance: Int): CubeCoordinates {
        return this.shift(CubeCoordinates(0, 0, 0), distance)
    }

    fun shift(start: CubeCoordinates, distance: Int = 1): CubeCoordinates {
        var shiftX = 0
        var shiftY = 0
        var shiftZ = 0
        when(this) {
            RIGHT -> {
                shiftX = start.x + distance
                shiftY = start.y - distance
            }
            LEFT -> {
                shiftX = start.x - distance
                shiftY = start.y + distance
            }
            UPRIGHT -> {
                shiftX = start.x + distance
                shiftZ = start.z - distance
            }
            UPLEFT -> {
                shiftY = start.y + distance
                shiftZ = start.z - distance
            }
            DOWNRIGHT -> {
                shiftY = start.y - distance
                shiftZ = start.z + distance
            }
            DOWNLEFT -> {
                shiftX = start.x - distance
                shiftZ = start.z + distance
            }
        }
        return CubeCoordinates(shiftX, shiftY, shiftZ)
    }
}
